### PR TITLE
fix: compile TS dependencies of JS files

### PR DESCRIPTION
**Pull request by bartlomieju**
Fixes https://github.com/denoland/deno/issues/5982

### CODE REVIEW

File: cli/global_state.rs [hunk 1]
```diff
@@ -141,13 +142,26 @@ impl GlobalState {
...
+    });
+
+    let needs_compilation = has_ts_dep
+      || (match out.media_type {
+        msg::MediaType::TypeScript
+        | msg::MediaType::TSX
+        | msg::MediaType::JSX => true,
+        msg::MediaType::JavaScript => self.ts_compiler.compile_js,
+        _ => false,
+      });
```
Comment (wcarss) [inline cli/global_state.rs:164]: drive-by commenter here; if the module graph may be large, wouldn't you rather run the simple match first?

Comment (bartlomieju) [reply cli/global_state.rs:164]: Doesn't really matter as it's the cheapest operation in whole compilation stack, but I refactored it to better reflect the intention

Comment (wcarss) [reply cli/global_state.rs]: thanks for having taken time to listen to a passerby on a low consequence suggestion :)

File: cli/tests/integration_tests.rs [hunk 2]
```diff
@@ -1837,6 +1837,11 @@ itest!(cjs_imports {
   output: "cjs_imports.ts.out",
 });
 
+itest!(ts_import_from_js {
+  args: "run --quiet --reload ts_import_from_js.js",
+  output: "ts_import_from_js.js.out",
+});
```
Comment (ry) [inline cli/tests/integration_tests.rs:1843]: This test already passes for me with v1.0.3 ...

File: cli/global_state.rs [hunk 3]
```diff
@@ -140,15 +141,27 @@ impl GlobalState {
...
       _ => false,
     };
 
+    needs_compilation |= module_graph.values().any(|module_file| {
+      let media_type = module_file.media_type;
+
+      media_type == (MediaType::TypeScript as i32)
+        || media_type == (MediaType::TSX as i32)
+        || media_type == (MediaType::JSX as i32)
+    });
```
Comment (ry) [inline cli/global_state.rs:163]: Consider factoring "needs_compilation" out into a separate utility function that can be unit tests independently.

File: cli/tests/integration_tests.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/tests/integration_tests.rs:1843]: That's strange; because it fails for me. Make sure to run with `--reload` flag - otherwise compiler might use already compiler file.

File: cli/global_state.rs (hunk 3 above)
Comment (bartlomieju) [reply cli/global_state.rs:163]: Sure

File: cli/tests/integration_tests.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/tests/integration_tests.rs:1843]: Changed the test case - it fails on current master and `1.0.3`

File: cli/global_state.rs [hunk 4]
```diff
@@ -241,8 +241,62 @@ impl GlobalState {
...
 #[test]
 fn thread_safe() {
   fn f<S: Send + Sync>(_: S) {}
   f(GlobalState::mock(vec![]));
 }
+
+#[test]
+fn test_needs_compilation() {
+  assert!(!needs_compilation(false, MediaType::JavaScript, vec![]));
+  assert!(needs_compilation(true, MediaType::JavaScript, vec![]));
```
Comment (ry) [inline cli/global_state.rs:281]: Could you add one more example?
```rust
assert!(!needs_compilation(true, TODO));
```
I'm not sure what TODO should be...

File: cli/tests/integration_tests.rs (hunk 2 above)
Comment (ry) [reply cli/tests/integration_tests.rs:1843]: I confirmed this new test fails on 1.0.3 👍

File: cli/global_state.rs (hunk 4 above)
Comment (bartlomieju) [reply cli/global_state.rs:298]: Done
